README.md
# loyalty-benefits-client

Loyalty Benefits API
- API version: 1.0.0
  - Build date: 2019-12-30T13:58:23.312-06:00[America/Chicago]

The Loyalty Benefits API provides benefits management including presentment and selection.  It also supports insurance certificate generation for insurance benefits that need certificates.


*Automatically generated by the [OpenAPI Generator](https://openapi-generator.tech)*


## Requirements

Building the API client library requires:
1. Java 1.7+
2. Maven/Gradle

## Installation

To install the API client library to your local Maven repository, simply execute:

```shell
mvn clean install
```

To deploy it to a remote Maven repository instead, configure the settings of the repository and execute:

```shell
mvn clean deploy
```

Refer to the [OSSRH Guide](http://central.sonatype.org/pages/ossrh-guide.html) for more information.

### Maven users

Add this dependency to your project's POM:

```xml
<dependency>
  <groupId>com.mastercard.developer</groupId>
  <artifactId>loyalty-benefits-client</artifactId>
  <version>1.0.0</version>
  <scope>compile</scope>
</dependency>
```

### Gradle users

Add this dependency to your project's build file:

```groovy
compile "com.mastercard.developer:loyalty-benefits-client:1.0.0"
```

### Others

At first generate the JAR by executing:

```shell
mvn clean package
```

Then manually install the following JARs:

* `target/loyalty-benefits-client-1.0.0.jar`
* `target/lib/*.jar`

## Getting Started

Please follow the [installation](#installation) instruction and execute the following Java code:

```java

import com.mastercard.developer.loyalty_benefits_client.*;
import com.mastercard.developer.loyalty_benefits_client.auth.*;
import com.mastercard.developer.loyalty_benefits_client.model.*;
import com.mastercard.developer.loyalty_benefits_client.api.BenefitsApi;

import java.io.File;
import java.util.*;

public class BenefitsApiExample {

    public static void main(String[] args) {
        
        BenefitsApi apiInstance = new BenefitsApi();
        Integer ica = 4567; // Integer | Card Issuer's ICA Number (Interbank Card Association number). A four digit number assigned by MasterCard to a financial institution.
        String userId = user1235; // String | Opaque identifier for the consumer. This is the same userId used while enrolling for benefits via Bundle Profile API.
        String panLastFourDigits = 2244; // String | Last four digits of user's registered 16 digit credit card number
        String channel = mobile; // String | Delivery channel.  One of mobile, tablet, or web
        String preferredLanguage = pt-BR; // String | User's preferred language in localized ISO 639-1 format such as pt-BR
        try {
            AssignedBenefits result = apiInstance.loyaltyBenefitsAssignedBenefitsGet(ica, userId, panLastFourDigits, channel, preferredLanguage);
            System.out.println(result);
        } catch (ApiException e) {
            System.err.println("Exception when calling BenefitsApi#loyaltyBenefitsAssignedBenefitsGet");
            e.printStackTrace();
        }
    }
}

```

## Documentation for API Endpoints

All URIs are relative to *https://api.mastercard.com*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*BenefitsApi* | [**loyaltyBenefitsAssignedBenefitsGet**](docs/BenefitsApi.md#loyaltyBenefitsAssignedBenefitsGet) | **GET** /loyalty/benefits/assigned-benefits | List of assigned benefits
*BenefitsApi* | [**loyaltyBenefitsBenefitSelectionsPost**](docs/BenefitsApi.md#loyaltyBenefitsBenefitSelectionsPost) | **POST** /loyalty/benefits/benefit-selections | Select optional benefits
*BenefitsApi* | [**loyaltyBenefitsProductBenefitsGet**](docs/BenefitsApi.md#loyaltyBenefitsProductBenefitsGet) | **GET** /loyalty/benefits/product-benefits | List all default and selectable benefits for the card product
*BundleProfileApi* | [**createUser**](docs/BundleProfileApi.md#createUser) | **POST** /bundle/profile/users | Create Profile
*InsuranceApi* | [**loyaltyInsuranceClaimSearchesPost**](docs/InsuranceApi.md#loyaltyInsuranceClaimSearchesPost) | **POST** /loyalty/insurance/claim-searches | Retrieve the claim status
*InsuranceApi* | [**loyaltyInsuranceUserDataTokensPost**](docs/InsuranceApi.md#loyaltyInsuranceUserDataTokensPost) | **POST** /loyalty/insurance/user-data-tokens | Create a user data token for submitting a new claim


## Documentation for Models

 - [Account](docs/Account.md)
 - [Address](docs/Address.md)
 - [Answers](docs/Answers.md)
 - [AssignedBenefit](docs/AssignedBenefit.md)
 - [AssignedBenefitCategory](docs/AssignedBenefitCategory.md)
 - [AssignedBenefitSubCategory](docs/AssignedBenefitSubCategory.md)
 - [AssignedBenefits](docs/AssignedBenefits.md)
 - [AssignedBenefitsBenefitSelectionWindow](docs/AssignedBenefitsBenefitSelectionWindow.md)
 - [Benefit](docs/Benefit.md)
 - [BenefitAttribute](docs/BenefitAttribute.md)
 - [BenefitCategory](docs/BenefitCategory.md)
 - [BenefitDetail](docs/BenefitDetail.md)
 - [BenefitDetailContent](docs/BenefitDetailContent.md)
 - [BenefitDetailContentPostselection](docs/BenefitDetailContentPostselection.md)
 - [BenefitDocuments](docs/BenefitDocuments.md)
 - [BenefitEligibilityQuestions](docs/BenefitEligibilityQuestions.md)
 - [BenefitSubCategory](docs/BenefitSubCategory.md)
 - [Benefits](docs/Benefits.md)
 - [BenefitsProgramTerms](docs/BenefitsProgramTerms.md)
 - [BenefitsSelection](docs/BenefitsSelection.md)
 - [BenefitsSelectionAdditionalInformation](docs/BenefitsSelectionAdditionalInformation.md)
 - [BilheteRequest](docs/BilheteRequest.md)
 - [BilheteRequestAdditionalInformation](docs/BilheteRequestAdditionalInformation.md)
 - [BilheteRequestAdditionalInformationCardholder](docs/BilheteRequestAdditionalInformationCardholder.md)
 - [BilheteRequestAdditionalInformationDependents](docs/BilheteRequestAdditionalInformationDependents.md)
 - [BilheteRequestAdditionalInformationPartner](docs/BilheteRequestAdditionalInformationPartner.md)
 - [BundleUser](docs/BundleUser.md)
 - [BundleUserData](docs/BundleUserData.md)
 - [BundleUserPatch](docs/BundleUserPatch.md)
 - [BundleUserResponse](docs/BundleUserResponse.md)
 - [BundleUserResponseData](docs/BundleUserResponseData.md)
 - [BundleUserResponseErrors](docs/BundleUserResponseErrors.md)
 - [ClaimFeature](docs/ClaimFeature.md)
 - [ClaimStatus](docs/ClaimStatus.md)
 - [ClaimStatusRequest](docs/ClaimStatusRequest.md)
 - [Consent](docs/Consent.md)
 - [EligibilityQuestions](docs/EligibilityQuestions.md)
 - [Email](docs/Email.md)
 - [Error](docs/Error.md)
 - [ErrorItem](docs/ErrorItem.md)
 - [ErrorItems](docs/ErrorItems.md)
 - [ErrorItemsErrors](docs/ErrorItemsErrors.md)
 - [Errors](docs/Errors.md)
 - [ErrorsErrors](docs/ErrorsErrors.md)
 - [Identification](docs/Identification.md)
 - [Im](docs/Im.md)
 - [IsRegistered](docs/IsRegistered.md)
 - [Name](docs/Name.md)
 - [PatchDocument](docs/PatchDocument.md)
 - [PhoneNumber](docs/PhoneNumber.md)
 - [Photo](docs/Photo.md)
 - [Questions](docs/Questions.md)
 - [RequiredClaimDocuments](docs/RequiredClaimDocuments.md)
 - [SubmitClaimUserData](docs/SubmitClaimUserData.md)
 - [SubmitClaimUserToken](docs/SubmitClaimUserToken.md)
 - [User](docs/User.md)
 - [UserProduct](docs/UserProduct.md)
 - [UserProductResponse](docs/UserProductResponse.md)
 - [UserReadAccountResponse](docs/UserReadAccountResponse.md)


## Documentation for Authorization

All endpoints do not require authorization.
Authentication schemes defined for the API:

## Recommendation

It's recommended to create an instance of `ApiClient` per thread in a multithreaded environment to avoid any potential issues.

## Author

loyalty-benefits-support@mastercard.flowdock.com

## Steps to run the application from command line


- Create a new project from Mastercard DevZone - stage.developer.mastercard.com or developer.mastercard.com


- Select "Loyalty Benefits" from Choose API dropdown and hit continue.


- Get Sandbox keys and store your .p12 certificate along with the readme/documentation.


- Please save this Sandbox Keys, .p12, key store password and alias as you are going to use these to run the application.


- Clone this repository and set up as Maven project


- Update the following keys in application.properties file.
	- *mastercard.benefits.ref.app.consumer.key*: This can be found in the project you created on developerZone
	- *mastercard.benefits.ref.app.keystore.path*: Path where you saved your certs i.e., .p12 file you received while creating a project
	- *mastercard.benefits.ref.app.keystore.password*: This is the password you get with Sandbox cert.
	- *mastercard.benefits.ref.app.keystore.alias*: This is the alias you get with Sandbox cert.


- Example:

```
	mastercard.benefits.ref.app.url = https://sandbox.api.mastercard.com
	mastercard.benefits.ref.app.consumer.key = Abcdfefgjhilklmnopqrstuvwxyz-dxcq_zD7IiPa0df175e!22a7fddba5
	mastercard.benefits.ref.app.keystore.path = C:\path\provided.p12
	mastercard.benefits.ref.app.keystore.password = pwd
	mastercard.benefits.ref.app.keystore.alias = alias
```


- Do a clean build either through IDE or command prompt, if you are doing it through command prompt then the below command should be executed in the directory which contains this repository's pom file Example: mvn clean install


- Run the application using this command: java -jar path of the Jar relative to the current directory/loyalty-benefits-client-1.0.0.jar


- Argument: An argument which defines the feature user wants to run through command line. If you don't specify this argument, it will run all the features(registration,getbenefits,selectbenefits,assignedbenefits and error) one after the other

registration : Registration for benefits service

getbenefits : Retrieve all available benefits

selectbenefits : Select Benefits to add to card holder

assignedbenefits : Retrieve benefits assigned to the user

error: An error scenario example

- Command line example to run one feature of the application:
`java -jar target/loyalty-benefits-client-1.0.0.jar getbenefits` using this command the application runs only getbenefits feature. 
 
- To run more than one feature then specify the features with comma separated. Example: `java -jar target/loyalty-benefits-client-1.0.0.jar registration,getbenefits,error` here it executes only these 3 features.

- To run all the features use this command: `java -jar target/loyalty-benefits-client-1.0.0.jar`. 

- To run the feature one after the other execute the command `java -jar target/loyalty-benefits-client-1.0.0.jar registration` then again run the command with different feature `java -jar target/loyalty-benefits-client-1.0.0.jar getbenefits` and so on.


## Reference App testing against Mock Service

- This app by default is compatible with the mock service of mastercard which only responds if it matches the exact request which the mock service expects (irrespective of required and optional fields).
- The word `reference` in the URL path indicates that it is going to test against the mock service, in order to build a client to test against the actual service remove the word `reference` in the path.
- You can expect the following response if the request doesn't match with expectation of the mock service.

```
<html>
    <head>
        <title>404 Not Found</title>
    </head>
    <body>
        <h1>Not Found</h1>
        <p>The requested URL was not found on this server.</p>
        <hr/>
        <p>
            <i>The DevTest VSE service could not match your request to a recorded request.&nbsp; Consider expanding your service image.</i>
        </p>
        <br/>
        <font size="-2">Produced by a DevTest virtualized web server.</font>
    </body>
</html>

```

## Testing in Sandbox environment

- To test this in Sandbox environment please contact Mastercard representative to set up you or your organization in this environment because if this does not happen the authorization fails in the Loyalty Benefits Service API. All the URLs have a prefix `reference` in this reference application for all resources of Loyalty Benefits Service API so that it deals with sample data. You need to remove this `reference` word from the URLs when testing against real data. Example: `/loyalty/benefits/reference` just for reference application and `/loyalty/benefits` to test against real data. 
    
 
-  [Generate API Clients](https://developer.mastercard.com/blog/consuming-mastercard-apis-in-client-applications) to consume Mastercard APIs.
Find the [service documentation](https://developer.mastercard.com/documentation/loyalty-benefits) for more information. 



- Client libraries can be generated for a simplified integration with the reference service, [for more details](https://developer.mastercard.com/blog/consuming-mastercard-apis-in-client-applications)


Copyright (c) 2019 Mastercard

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License

CHANGELOG.md
## 1.0.0

- A reference implementation to use Loyalty Benefits services API.

src/main/java/com/mastercard/developer/example/BenefitsApiDemo.java
/*Copyright (c) 2019 Mastercard
 
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 
    http://www.apache.org/licenses/LICENSE-2.0
 
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

*Loyalty Benefits API
 * The Loyalty Benefits API provides benefits management including presentment and selection.  It also supports insurance certificate generation for insurance benefits that need certificates.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: loyalty-benefits-support@mastercard.flowdock.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.*/
package com.mastercard.developer.example;

import java.util.ArrayList;
import java.util.List;

import com.mastercard.developer.interceptors.OkHttp2OAuth1Interceptor;
import com.mastercard.developer.loyalty_benefits_client.ApiClient;
import com.mastercard.developer.loyalty_benefits_client.ApiException;
import com.mastercard.developer.loyalty_benefits_client.api.BenefitsApi;
import com.mastercard.developer.loyalty_benefits_client.api.BundleProfileApi;
import com.mastercard.developer.loyalty_benefits_client.api.InsuranceApi;
import com.mastercard.developer.loyalty_benefits_client.model.Account;
import com.mastercard.developer.loyalty_benefits_client.model.AccountExternal;
import com.mastercard.developer.loyalty_benefits_client.model.Address;
import com.mastercard.developer.loyalty_benefits_client.model.AssignedBenefits;
import com.mastercard.developer.loyalty_benefits_client.model.Benefits;
import com.mastercard.developer.loyalty_benefits_client.model.BenefitsSelection;
import com.mastercard.developer.loyalty_benefits_client.model.BenefitsSelectionAdditionalInformation;
import com.mastercard.developer.loyalty_benefits_client.model.BilheteRequestAdditionalInformationCardholder;
import com.mastercard.developer.loyalty_benefits_client.model.BilheteRequestAdditionalInformationDependents;
import com.mastercard.developer.loyalty_benefits_client.model.BilheteRequestAdditionalInformationPartner;
import com.mastercard.developer.loyalty_benefits_client.model.BundleUser;
import com.mastercard.developer.loyalty_benefits_client.model.BundleUserData;
import com.mastercard.developer.loyalty_benefits_client.model.BundleUserResponse;
import com.mastercard.developer.loyalty_benefits_client.model.ClaimStatus;
import com.mastercard.developer.loyalty_benefits_client.model.ClaimStatusRequest;
import com.mastercard.developer.loyalty_benefits_client.model.Consent;
import com.mastercard.developer.loyalty_benefits_client.model.Credentials;
import com.mastercard.developer.loyalty_benefits_client.model.Email;
import com.mastercard.developer.loyalty_benefits_client.model.Identification;
import com.mastercard.developer.loyalty_benefits_client.model.Im;
import com.mastercard.developer.loyalty_benefits_client.model.Name;
import com.mastercard.developer.loyalty_benefits_client.model.PhoneNumber;
import com.mastercard.developer.loyalty_benefits_client.model.Photo;
import com.mastercard.developer.loyalty_benefits_client.model.SubmitClaimUserData;
import com.mastercard.developer.loyalty_benefits_client.model.SubmitClaimUserToken;
import com.mastercard.developer.loyalty_benefits_client.model.User;
import com.mastercard.developer.loyalty_benefits_client.model.UserProduct;

public class BenefitsApiDemo {

	public static void main(String[] args) throws Exception {

		RequestHelper.loadProperties();

		ApiClient apiClient = new ApiClient();
		apiClient.setBasePath(RequestHelper.getBaseURL());
		apiClient.getHttpClient().interceptors()
				.add(new OkHttp2OAuth1Interceptor(RequestHelper.getConsumerkey(), RequestHelper.getPrivateKey()));
		apiClient.setDebugging(true);
		
		BenefitsApi benefitsApi = new BenefitsApi(apiClient);

		InsuranceApi insuranceApi = new InsuranceApi(apiClient);
		
		if(runThisScenario(args,"registration")) {
			BundleProfileApi bundleProfileApi = new BundleProfileApi(apiClient);
			executeRegistrationScenario(bundleProfileApi);
		}
		
		if(runThisScenario(args,"getbenefits")) {
			executeBenefitsRetrieveScenario(benefitsApi);
		}

		if(runThisScenario(args,"selectbenefits")) {
			executeBenefitSelectionScenario(benefitsApi);
		}

		if(runThisScenario(args,"assignedbenefits")) {
			executeAssignedBenefitsRetrieveScenario(benefitsApi);
		}

		if(runThisScenario(args,"claimsearch")) {
			executeInsuranceClaimSearchScenario(insuranceApi);
		}
		
		if(runThisScenario(args,"insurancetoken")) {
			executeInsuranceTokenDataScenario(insuranceApi);
		}

		if(runThisScenario(args,"error")) {
			executeBenefitsRetrieveErrorScenario(benefitsApi);
		}
	}

	private static void executeRegistrationScenario(BundleProfileApi bundleProfileApi) {
		BundleUser bundleUser = getRegistrationObject();
		try {
			printMessage("STARTING BENEFITS REGISTRATION FROM COMMAND LINE RUNNER");
			BundleUserResponse result = bundleProfileApi.createUser(bundleUser);
			System.out.println("Parsed Response---------------");
			System.out.println(result);
		} catch (Exception e) {
			System.err.println("Exception when calling Benefits Registration");
			e.printStackTrace();
		}
	}

	private static void executeBenefitsRetrieveScenario(BenefitsApi benefitsApi) {
		Integer ica = 1216;
		String productName = "black";
		String channel = "web";
		String preferredLanguage = "pt-BR";
		try {
			printMessage("STARTING GET AVAILABLE BENEFITS FROM COMMAND LINE RUNNER");
			Benefits result = benefitsApi.loyaltyBenefitsProductBenefitsGet(ica, productName, channel,
					preferredLanguage);
			System.out.println("Parsed Response---------------");
			System.out.println(result);
		} catch (Exception e) {
			System.err.println("Exception when calling Get Benefits");
			e.printStackTrace();
		}
	}

	private static void executeBenefitSelectionScenario(BenefitsApi benefitsApi) {
		BenefitsSelection benefitsSelection = getBenefitSelectionObject();
		try {
			printMessage("STARTING BENEFITS SELECTION FROM COMMAND LINE RUNNER");
			benefitsApi.loyaltyBenefitsBenefitSelectionsPost(benefitsSelection);
			System.out.println("Benefit Selection Successfull");
		} catch (Exception e) {
			System.err.println("Exception while Benefit Selection");
			e.printStackTrace();
		}
	}

	private static void executeAssignedBenefitsRetrieveScenario(BenefitsApi benefitsApi) {
		Integer ica = 1216;
		String panLastFourDigits = "2244";
		String channel = "tablet";
		String preferredLanguage = "en-GB";
		String userId = "user1235";
		try {
			printMessage("STARTING GET ASSIGNED BENEFITS FROM COMMAND LINE RUNNER");
			AssignedBenefits result = benefitsApi.loyaltyBenefitsAssignedBenefitsGet(ica, userId, panLastFourDigits,
					channel, preferredLanguage);
			System.out.println("Parsed Response---------------");
			System.out.println(result);
		} catch (Exception e) {
			System.err.println("Exception when calling Assigned Benefits");
			e.printStackTrace();
		}
	}

	private static void executeInsuranceClaimSearchScenario(InsuranceApi insuranceApi) {
		ClaimStatusRequest claimStatusRequest = new ClaimStatusRequest("C0011", "71000", "English");
		try {
			printMessage("STARTING INSURANCE CLAIM SEARCH FROM COMMAND LINE RUNNER");
			ClaimStatus result = insuranceApi.loyaltyInsuranceClaimSearchesPost(claimStatusRequest);
			System.out.println("Parsed Response---------------");
			System.out.println(result);
		} catch (Exception e) {
			System.err.println("Exception in Insurance Claim Search");
			e.printStackTrace();
		}
	}
	
	private static void executeInsuranceTokenDataScenario(InsuranceApi insuranceApi) {
		
		SubmitClaimUserData submitClaimUserData = new SubmitClaimUserData();
		submitClaimUserData.setFirstName("John");
		submitClaimUserData.setLastName("Smith");
		submitClaimUserData.setState("SP");
		submitClaimUserData.setPan("50000000000000000");

		try {
			printMessage("STARTING INSURANCE TOKENIZE DATA FROM COMMAND LINE RUNNER");
			SubmitClaimUserToken result = insuranceApi.loyaltyInsuranceUserDataTokensPost(submitClaimUserData);
			System.out.println("Parsed Response---------------");
			System.out.println("Token: "+result);
		} catch (Exception e) {
			System.err.println("Exception in Insurance Token Data");
			e.printStackTrace();
		}
	}
	
	private static void executeBenefitsRetrieveErrorScenario(BenefitsApi benefitsApi) {
		Integer ica = 1;
		String productName = "KWrkp2V4c";
		String channel = "rpBLflhhahT";
		String preferredLanguage = "Mvr2Avshto";
		try {
			printMessage("STARTING GET AVAILABLE BENEFITS ERROR SCENARIO FROM COMMAND LINE RUNNER");
			benefitsApi.loyaltyBenefitsProductBenefitsGet(ica, productName, channel,
					preferredLanguage);
		} catch (ApiException e) {
			System.err.println("Exception when calling BenefitsApi#GetBenefits");
			System.out.println("Parsed Response---------------");
			System.out.println(e.getResponseBody());
		}
	}

	private static void printMessage(String scenario) {
		System.out.println("--------------------------------------------------------------------");
		System.out.println("--------------------------------------------------------------------");
		System.out.println(scenario + "---------------");
		System.out.println("--------------------------------------------------------------------");
		System.out.println("--------------------------------------------------------------------");
	}

	private static BenefitsSelection getBenefitSelectionObject() {

		BenefitsSelection benefitsSelection = new BenefitsSelection();
		benefitsSelection.setUserId("user1235");
		benefitsSelection.setPanLastFourDigits("2244");
		benefitsSelection.setIca(1216);
		benefitsSelection.setCardProductType("Black");
		benefitsSelection.setMemberId("3154");

		List<String> benefitIds = new ArrayList<>();
		benefitIds.add("BEN7890");
		benefitsSelection.setBenefitIds(benefitIds);

		BenefitsSelectionAdditionalInformation additionalInformation = new BenefitsSelectionAdditionalInformation();
		additionalInformation.setCardholder(new BilheteRequestAdditionalInformationCardholder("John", "Smith",
				"1992-02-02", "BRA", "999.999.999-99", "Rua Francisco Pereira da Silva 1480", "Guarulhos", "SP",
				"1930288", "Parque Paulistano", "john.smith@mail.com", "11", "1111-1111"));
		additionalInformation.setPartner(
				new BilheteRequestAdditionalInformationPartner("John", "Smith", "999.999.999-99", "1992-02-02"));
		List<BilheteRequestAdditionalInformationDependents> dependents = new ArrayList<>();
		dependents.add(
				new BilheteRequestAdditionalInformationDependents("John", "Smith", "999.999.999-99", "1991-01-02"));
		additionalInformation.setDependents(dependents);
		benefitsSelection.setAdditionalInformation(getAdditionalInformation());

		return benefitsSelection;

	}

	private static BenefitsSelectionAdditionalInformation getAdditionalInformation() {
		BenefitsSelectionAdditionalInformation additionalInformation = new BenefitsSelectionAdditionalInformation();
		additionalInformation.setCardholder(new BilheteRequestAdditionalInformationCardholder("John", "Smith",
				"1992-02-02", "BRA", "999.999.999-99", "Rua Francisco Pereira da Silva 1480", "Guarulhos", "SP",
				"1930288", "Parque Paulistano", "john.smith@mail.com", "11", "1111-1111"));
		additionalInformation.setPartner(
				new BilheteRequestAdditionalInformationPartner("John", "Smith", "999.999.999-99", "1992-02-02"));
		List<BilheteRequestAdditionalInformationDependents> dependents = new ArrayList<>();
		dependents.add(
				new BilheteRequestAdditionalInformationDependents("John", "Smith", "999.999.999-99", "1991-01-02"));
		additionalInformation.setDependents(dependents);
		return additionalInformation;
	}

	private static BundleUser getRegistrationObject() {
		BundleUser bundleUser = new BundleUser();
		BundleUserData bundleUserData = new BundleUserData();
		User user = new User();
		UserProduct userProduct = new UserProduct();
		Address address = new Address();
		Email email = new Email();
		Identification identification = new Identification();
		Im im = new Im();
		Name name = new Name();
		PhoneNumber phoneNumber = new PhoneNumber();
		Photo photo = new Photo();
		List<Photo> photos = new ArrayList<>();
		List<PhoneNumber> phoneNumbers = new ArrayList<>();
		List<Im> ims = new ArrayList<>();
		List<Identification> identifications = new ArrayList<>();
		List<Email> emails = new ArrayList<>();
		List<Address> addresses = new ArrayList<>();

		user.setActive(true);
		address.setCountry("US");
		address.setFormatted("114 5th Ave, New York, NY 10011");
		address.setLocality("city");
		address.setOperation("None");
		address.setPostalCode("10011");
		address.setPrimary(false);
		address.setRegion("NY");
		address.setStreetAddress("114 5th Ave");
		address.setStreetAddress2("Suite 20");
		address.setType("work");
		addresses.add(address);
		user.setAddresses(addresses);

		user.setDateOfBirth("1992-11-17");
		user.setDisplayName("John F Smith");

		email.setDisplay("Primary email address of user");
		email.setOperation("None");
		email.setPrimary(true);
		email.setType("home");
		email.setValue("john5033379289074369@gmail.com");
		emails.add(email);
		user.setEmails(emails);

		user.setExternalId("user123-partnerBank");
		user.setId("d3459481-fb1c-48bb-8685-40eb629e2ae1");

		identification.setCountry("BR");
		identification.setType("CPF");
		identification.setValue("496881234");
		identifications.add(identification);
		user.setIdentifications(identifications);

		im.setDisplay("user Yahoo-id");
		im.setOperation("None");
		im.setPrimary(true);
		im.setType("aim");
		im.setValue("JohnFSmith123");
		ims.add(im);
		user.setIms(ims);

		user.locale("en-US");

		name.setFamilyName("Smitherines");
		name.setFormatted("Sally Smitherines");
		name.setGivenName("Sally");
		name.setHonorificPrefix("Ms.");
		name.setHonorificSuffix("Sr.");
		name.setMiddleName("F");
		user.setName(name);

		user.setNickName("sallySmith");
		user.setPassword("boingouser5105363487498185");

		phoneNumber.setDisplay("Office fax of user");
		phoneNumber.setOperation("None");
		phoneNumber.setPrimary(false);
		phoneNumber.setType("fax");
		phoneNumber.setValue("+1-201-555-0123");
		phoneNumbers.add(phoneNumber);
		user.setPhoneNumbers(phoneNumbers);

		photo.setDisplay("portrait");
		photo.setOperation("None");
		photo.setPrimary(true);
		photo.setType("thumbnail");
		photo.setValue("facebook.com/JohnFSmith123/photos/1");
		photos.add(photo);
		user.setPhotos(photos);

		user.setPreferredLanguage("en");
		user.setProfileUrl("facebook.com/JohnFSmith123");
		user.setTimezone("America/New_York");
		user.setTitle("Vice President");
		user.setUserId("user123-partnerBank");
		user.setUserName("JohnFSmith123");
		user.setUserType("Intern");

		Account account = new Account();
		account.setAccountType("SingleMember");
		account.setBrand("Mastercard");
		account.setCardAlias("d3459481-fb1c-48bb-8685-40eb629e2ae1");
		account.setCardExpiryDate("02/2024");
		account.setCvcCode("876");
		account.setIca(12594L);
		account.setLast4Pan(8185);
		account.setNameOnCard("SALLY SMITHERINES");
		account.setPan("5105363487498185");
		account.setPar("q1hjz28rka1ebl470g9xyg90r5d3e");
		account.setPrimary(true);
		account.setProductLine("Credit cards");
		account.setProductType("Black");
		account.setUuid("d3459481-fb1c-48bb-8685-40eb629e2ae1");

		AccountExternal accountExternal = new AccountExternal();
		accountExternal.setAccountStatusCode("1");
		accountExternal.setExternalMembershipReferenceId("B12345");
		account.setObject(accountExternal);

		Consent consent = new Consent();
		consent.setAgreed(true);
		consent.setConsentType("author-legal-content-document");
		consent.setConsentedDate("2018-06-19T12:30:42.307+0000");
		consent.setCountry("BRA");
		consent.setLanguage("en_us");
		consent.setServiceCode("priceless");
		consent.setServiceFunctionCode("registration-page code");
		consent.setUseCategoryCode("UC-01");
		consent.setUseCategoryValue("acceptPersonalizedOffersEmail");
		consent.setUuid("d3459481-fb1c-48bb-8685-40eb629e2ae1");

		List<Account> accounts = new ArrayList<>();
		accounts.add(account);
		List<Consent> consents = new ArrayList<>();
		consents.add(consent);
		Credentials productCredentials = new Credentials();
		productCredentials.setUsername("boingoid123");
		productCredentials.setPassword("boingopwd");
		userProduct.setProduct("benefits");
		userProduct.setAccounts(accounts);
		userProduct.setConsents(consents);
		userProduct.setObject(productCredentials);
		List<UserProduct> products = new ArrayList<>();
		products.add(userProduct);

		bundleUserData.setProducts(products);
		bundleUserData.setUser(user);
		bundleUser.setData(bundleUserData);

		return bundleUser;

	}

	private static boolean runThisScenario(String[] args, String scenario) {
		return (args!=null && args.length>0 && args[0].contains(scenario)) || (args == null || args.length == 0);
	}
}
